### PR TITLE
add the catalog endpoint

### PR DESCRIPTION
**Pull request by nimakaviani**
- make the service configurable through the service file
- parse configuration into brokerapi.Service
- refactor Config and service api endpoints

finishes #5

### CODE REVIEW

File: pkg/broker/broker_test.go [hunk 1]
```diff
@@ -4,17 +4,31 @@ import (
 	"context"
 
 	. "github.com/cloudfoundry-incubator/blockhead/pkg/broker"
+	"github.com/cloudfoundry-incubator/blockhead/pkg/config"
```
Comment (MHBauer) [inline pkg/broker/broker_test.go]: fix this. dot import or import normally, not both.

File: pkg/broker/broker.go [hunk 2]
```diff
@@ -3,36 +3,46 @@ package broker
...
+		b.config.Service,
+	}, nil
+}
+
+func (b BlockheadBroker) Provision(ctx context.Context, instanceID string, details brokerapi.ProvisionDetails, asyncAllowed bool) (brokerapi.ProvisionedServiceSpec, error) {
 	return brokerapi.ProvisionedServiceSpec{}, nil
 }
 
-func (blockheadbroker BlockheadBroker) Deprovision(ctx context.Context, instanceID string, details brokerapi.DeprovisionDetails, asyncAllowed bool) (brokerapi.DeprovisionServiceSpec, error) {
+func (b BlockheadBroker) Deprovision(ctx context.Context, instanceID string, details brokerapi.DeprovisionDetails, asyncAllowed bool) (brokerapi.DeprovisionServiceSpec, error) {
```
Comment (MHBauer) [inline pkg/broker/broker.go]: separate out random renames into a separate commit.

File: Dockerfile [hunk 3]
```diff
@@ -6,5 +6,5 @@ run go build -v -o /broker ./cmd/broker
 from alpine:3.8
 expose 3333
 copy --from=builder /broker /broker
-add ./creds.json /creds.json
-cmd ["/broker","/creds.json"]
+add ./config.json /config.json
+cmd ["/broker","/config.json"]
```
Comment (MHBauer) [inline Dockerfile]: separate out random renames into a separate commit.

File: pkg/config/config.go [hunk 4]
```diff
@@ -3,16 +3,18 @@ package config
...
 	"encoding/json"
 	"fmt"
+	"github.com/pivotal-cf/brokerapi"
 	"io/ioutil"
 )
 
-type BlockheadConfig struct {
-	Password string `yaml:"password,omitempty"`
-	Username string `yaml:"username,omitempty"`
+type Config struct {
```
Comment (MHBauer) [inline pkg/config/config.go]: why this rename?

**Comment by jberkhahn** [comment]
We think it would make more sense to have the info describing the services be in separate files, similar to how the old ruby broker worked. It would make it much easier to maintain that functionality, both when using the broker stand alone and with the bosh-release. Maybe have the config contain the location of a directory to load service jsons from.

File: pkg/broker/broker_test.go (hunk 1 above)
Comment (MHBauer) [reply pkg/broker/broker_test.go]: derp, 
read too quickly, 
fine as is. 
somewhat concerned by the dot import at all tho.

**Comment by jberkhahn** [comment]
What do we need flags for? You could just make a field in the config that is the location of a dir to load the service files from.

**Comment by nimakaviani** [comment]
oh I was thinking `./broker -config [path] -service [path]`

relative / absolute path in the config file is kinda ugly.

**Comment by swetharepakula** [comment]
How would the `-service` flag work with multiple services? I like the idea of a directory or just providing a list of files. Unless the flag was to signal the array of filenames.

**Comment by nimakaviani** [comment]
generally you can use flags to path a slice through. like the following:

`./broker -config [path] -service [path1] -service [path2]`

e.g., similar to how ops files are passed through to a bosh cli.

If a directory, it should be internal and we should just read everything in it, rather than making it part of the config but that is implicit logic which is hard to reason about. In case of the broker bosh release, those `yml` files get passed into the deployment as bosh opsfiles .

File: pkg/config/config.go [hunk 5]
```diff
@@ -3,16 +3,18 @@ package config
...
+	Password string            `json:"password,omitempty"`
+	Username string            `json:"username,omitempty"`
+	Service  brokerapi.Service `json:"service"`
```
Comment (swetharepakula) [inline pkg/config/config.go]: This should be plural.

File: pkg/config/config_test.go [hunk 6]
```diff
@@ -8,19 +8,46 @@ import (
...
+			Expect(cfg.Service.Bindable).To(BeTrue())
+			Expect(cfg.Service.Tags).To(ConsistOf("eth", "geth"))
+			Expect(cfg.Service.Metadata.DisplayName).To(Equal("some-name"))
+			Expect(cfg.Service.Metadata.LongDescription).To(Equal("some-long-desc"))
+			Expect(cfg.Service.Metadata.ProviderDisplayName).To(Equal("some-provider-display-name"))
+			Expect(cfg.Service.DashboardClient.ID).To(Equal("some-client-id"))
+			Expect(cfg.Service.DashboardClient.Secret).To(Equal("some-secret"))
+			Expect(cfg.Service.Plans).To(HaveLen(1))
+			Expect(cfg.Service.Plans[0].ID).To(Equal("some-plan-id"))
+			Expect(cfg.Service.Plans[0].Name).To(Equal("free"))
```
Comment (swetharepakula) [inline pkg/config/config_test.go]: Change this comparing to the expected struct, instead of checking each member line by line

File: pkg/config/config.go (hunk 5 above)
Comment (nimakaviani) [reply pkg/config/config.go]: why? I thought about making it plural but for the beginning we are going to only support eth in dev mode and in the best case with some plans for it. why do we need to make it more generic than that where we dont need to atm.

File: pkg/config/config_test.go (hunk 6 above)
Comment (nimakaviani) [reply pkg/config/config_test.go]: sounds good.

File: pkg/config/config.go (hunk 5 above)
Comment (swetharepakula) [reply pkg/config/config.go]: Because we know we want to support Fabric in the nearish future. If we had no plans I would agree but since we do have one making it work for multiple services makes more sense to me. This also ties into having the array of services in running the broker.

Comment (nimakaviani) [reply pkg/config/config.go]: point taken. A broker with multiple services probably makes for a good demo, but practically speaking for the case of Fabric I wonder if a single broker would be used to offer both the eth service and the fabric service. 

Anyhow making it a slice sounds good.

File: pkg/broker/broker_test.go [hunk 7]
```diff
@@ -57,36 +58,64 @@ var _ = Describe("Broker", func() {
...
+						},
+					},
+				},
+			}
 		})
 
 		It("should return service definition", func() {
 			services, err := blockhead.Services(ctx)
 			Expect(err).NotTo(HaveOccurred())
-			Expect(services).To(HaveLen(1))
```
Comment (swetharepakula) [inline pkg/broker/broker_test.go]: Is it possible to not add this in the previous commit and have the removal in this commit?

Comment (nimakaviani) [reply pkg/broker/broker_test.go]: yeah. honestly I am disliking this idea of having multiple commits under one PR. makes it super confusing and hard to review.

File: pkg/config/assets/service_config.json [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+       {
+          "id": "some-plan-id",
+          "name": "free",
+          "description": "free-trial",
+          "free": true,
+          "metadata": {
+             "displayName": "service-plan-metadata",
+             "container": {
+               "backend": "docker",
+               "image": "some-image"
```
Comment (swetharepakula) [inline pkg/config/assets/service_config.json]: Do we want to have this metadata on the service config or something the broker knows?
As in I think this is information for the broker not necessarily for the catalog?

Comment (nimakaviani) [reply pkg/config/assets/service_config.json]: I agree in that it probably doesnt need to be exposed as part of the catalog. but we still want to tie it to a service definition config somehow in such a way that a given variation of the service can use a different docker repo/image.

Comment (swetharepakula) [reply pkg/config/assets/service_config.json]: yea i agree with that, then we should probably test that this data does not show up in the service definition.

Comment (nimakaviani) [reply pkg/config/assets/service_config.json]: SGTM

**Comment by nimakaviani** [comment]
closed in favor of #23
